From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Roman Leventov <leventov.ru@gmail.com>
Date: Thu, 27 Aug 2026 13:49:20 +0800
Subject: [PATCH 1/5] Support: run serial tasks inline on single-threaded
 Emscripten

Avoid creating the executor thread when Emscripten is built without pthreads. Running the task inline preserves SerialExecutor ordering while allowing the single-threaded runtime to complete finalizers.
---
 lib/Support/SerialExecutor.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/Support/SerialExecutor.cpp b/lib/Support/SerialExecutor.cpp
index bd9b4788cd69d88a84350375baf060a988b79cbc..b60727994aaa33972021fda861d4d00c75720b43 100644
--- a/lib/Support/SerialExecutor.cpp
+++ b/lib/Support/SerialExecutor.cpp
@@ -32,6 +32,11 @@ SerialExecutor::~SerialExecutor() {
 }
 
 void SerialExecutor::add(std::function<void()> task) {
+#if defined(__EMSCRIPTEN__) && !defined(__EMSCRIPTEN_PTHREADS__)
+  task();
+  return;
+#endif
+
   std::unique_lock<std::mutex> lock(mutex_);
   assert(
       threadState_ != ThreadState::Terminating &&
